core/src/wsv.rs

```diff
@@ -54,7 +55,37 @@ pub struct World {
     /// Triggers
     pub(crate) triggers: TriggerSet,
     /// Chain of *runtime* validators
-    pub(crate) validator: crate::validator::Validator,
+    pub(crate) validator: RwLock<Option<crate::validator::Validator>>,
```

Comment:
as far as I am aware we will not need internal syncing inside `wsv`. The whole `wsv` will be guarded with a sync primitive

Reply:
I'd consider an atomic pointer. Semantically a `Mutex` (or `RWLock`) is useful when you're going to modify a piece of a data structure and not replace it completely. Here the most likely mode of modification is loading a whole different validator. 

This is not a huge deal, but with an atomic pointer, the synchronisation overhead is a single atomic read for access, and a single atomic write for updates, similar to an `RWLock`.

Reply:
@appetrosyan , I don't quite understand how `AtomicPtr` will help here, because we still need to store `WSV` somewhere and also it will require to use `unsafe` to deref the pointer.

Also I don't know what to do with this conversation now. Should we move it to another issue?

Reply:
Re: conversation — yes.

Reply:
> require `unsafe` to deref the pointer. 

[Not necessarily](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicPtr.html#method.get_mut). 

The `AtomicPtr` is useful when you are going to be replacing one struct with a new struct every time you update. Reading an atomic pointer is the same as reading a regular pointer. 

There's no locking, instead at the moment when your new struct instance is ready, a single atomic exchange replaces the old memory location with the location of the newly allocated struct. This is not a common pattern, primarily because re-allocation is more expensive than the locking in `Mutex` and `RWLock`. 

Here you know for certain that the validator is going to be updated rarely and usually with a new allocation. So you also know that removing the overhead of `lock()` is going to help. The only problem that could arise, is what to do if the validator is outdated. That question is solved by consensus. So you have nothing to lose and everything to gain.

Reply:
Created #3390. I'm closing this